In [1]:
import numpy as np
import matplotlib.pyplot as plt
plt.rcParams.update({'font.size': 14,
                     'xtick.labelsize': 14})


In [2]:
X = np.array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1],   # для умножения на intercept
              [1, 1, 2, 1, 3, 0, 5, 10, 1, 2]]).T # стаж репетитора
X.shape

(10, 2)

In [3]:
# средний балл ЕГЭ (целевая переменная)
y = [45, 55, 50, 59, 65, 35, 75, 80, 50, 60]

In [4]:
def calc_mse(y, y_pred):
    err = np.mean((y - y_pred)**2)
    return err

def calc_mae(y, y_pred):
    err = np.mean(np.abs(y - y_pred))
    return err

1. Подберите скорость обучения (alpha) и количество итераций для градиентного спуска (до совпадения ответов с результатами МНК). Как влияют друг на друга эти два параметра?

In [5]:
n = X.shape[0]
alpha = 1e-3
w = np.array([1, 0.5])

for i in range(1000+1):
    y_pred = np.dot(w, X.T)
    err = calc_mse(y, y_pred)
    for j in range(w.shape[0]):
        w[j] -= alpha * (1/n * 2 * np.sum(X[:,j] * (y_pred - y)))
    if i % 100 == 0:
        print(i, w, err) # МНК: array([47.23214286,  3.91071429]) 45.937499999999986

0 [1.1102 0.84  ] 3173.15
100 [ 7.43515582 10.71754582] 899.5990444359252
200 [11.44993589 10.49282195] 735.5540619653672
300 [14.98921624  9.8634316 ] 606.0048772266812
400 [18.17511588  9.27626113] 500.7989288428639
500 [21.04606457  8.74617136] 415.355423398327
600 [23.63334499  8.26841272] 345.9620646005472
700 [25.96499229  7.83785632] 289.6039104863001
800 [28.06626432  7.44984037] 243.83236130353217
900 [29.95992301  7.10016181] 206.65877484533297
1000 [31.66648131  6.78503276] 176.46806566379882


In [6]:
n = X.shape[0]
alpha = 5e-2
w = np.array([1, 0.5])

for i in range(501):
    y_pred = np.dot(w, X.T)
    err = calc_mse(y, y_pred)
    for j in range(w.shape[0]):
        w[j] -= alpha * (1/n * 2 * np.sum(X[:,j] * (y_pred - y)))
    if i % 100 == 0:
        print(i, w, err) # МНК: array([47.23214286,  3.91071429]) 45.937499999999986

0 [ 6.51 17.5 ] 3173.15
100 [47.03142979  3.94777749] 45.96159894380851
200 [47.23117922  3.91089223] 45.93750055548498
300 [47.23213823  3.91071514] 45.93750000001278
400 [47.23214283  3.91071429] 45.93749999999999
500 [47.23214286  3.91071429] 45.937499999999986


 Вообще при  скорость обучения (alpha) 5e-2, уже при 201 итераций получаем 200 [47.23117922  3.91089223] 45.93750055548498
Но если ставим 501 итерацию, то 500 [47.23214286  3.91071429] 45.937499999999986


*2. В этом коде мы избавляемся от итераций по весам, но тут есть ошибка, исправьте ее:

In [7]:
w = np.array([1, 0.5])

for i in range(501):
    y_pred = np.dot(w, X.T)
    err = calc_mse(y, y_pred)
    '''for j in range(w.shape[0]):
        w[j] -= alpha * (1/n * 2 * np.sum(X[j] * (y_pred - y)))'''
    w -= (alpha * (1/n * 2 * np.sum(X.T * (y_pred - y), axis=1)))
    if i % 100 == 0:
        print(i, w, err) # [47.23214286  3.91071429] 45.937499999999986

0 [ 6.51 17.5 ] 3173.15
100 [47.03142979  3.94777749] 45.96159894380851
200 [47.23117922  3.91089223] 45.93750055548498
300 [47.23213823  3.91071514] 45.93750000001278
400 [47.23214283  3.91071429] 45.93749999999999
500 [47.23214286  3.91071429] 45.937499999999986


Если добавим axis=1, то все получается

*3. Вместо того, чтобы задавать количество итераций, задайте условие остановки алгоритма - когда ошибка за итерацию начинает изменяться ниже определенного порога.

Сколько нужно сделать итераций, если установить допустимое отклонение mse в размере  diff=10−6 , а значение  𝛼=10−2 ?

In [12]:
# если нам надо остановить по разнице ошибок
n = X.shape[0]
alpha = 5e-2
w = np.array([1, 0.5])
err = float('Inf')
i =0

while True:
    i += 1
    y_pred = np.dot(w, X.T)
    new_err = calc_mse(y, y_pred)
    if err - new_err == 0:
        w -= (alpha * (1/n * 2 * np.sum(X.T * (y_pred - y), axis=1)))
        break
    else:    
        err = calc_mse(y, y_pred)
        w -= (alpha * (1/n * 2 * np.sum(X.T * (y_pred - y), axis=1)))
       

print(w, err)
print(f"Количество итераций {i}")

[47.23214237  3.91071438] 45.93750000000014
Количество итераций 343
